builtin/add.c

```diff
@@ -517,9 +519,15 @@ int cmd_add(int argc, const char **argv, const char *prefix)
 		}
 
 		/* This picks up the paths that are not tracked */
+		trace2_region_enter("add", "fill_directory", NULL);
 		baselen = fill_directory(&dir, &the_index, &pathspec);
-		if (pathspec.nr)
+		trace2_region_leave("add", "fill_directory", NULL);
+
+		if (pathspec.nr) {
+			trace2_region_enter("add", "prune_directory", NULL);
 			seen = prune_directory(&dir, &pathspec, baselen);
+			trace2_region_leave("add", "prune_directory", NULL);
+		}
```

Comment:
Leaving a comment here, since it's closest to the unmodified `if (pathspec.nr)` below.

Any chance that the `git add .` populates `pathspec` with one entry, `.`? Could it be spending any time in that section?

(If you have already tested with the current regions and they take into account most of the time, then feel free to ignore this suggestion.)

Reply:
> Any chance that the git add . populates pathspec with one entry, .?

Yep it does, and that actually causes problems in `validate_untracked_cache` that I've been working on sorting out because `validate_untracked_cache` is thinking that a filter is set and hence the untracked cache can't be used.  I have some local changes I've been trying out to address that issue.

> If you have already tested with the current regions and they take into account most of the time, then feel free to ignore this suggestion.

From what I've seen so far almost all of the time is being spent in `fill_directory` and `add_files_to_cache`.  I've been starting with `fill_directory` as I've confirmed that `git status` also calls it and is able to successfully  use the untracked cache.